✨Added displayMode prop

anto90fg opened the pull request:
#### What problem is this solving?

For SEO reason, there may be a need to have the link to the current product associated with the product name. For this reason, I have added the possibility of using a settable prop to include the product name in a link to the product page.

#### How to test it?

I insert a direct link to the plp, in which the component has been set and the prop set to true.

[Workspace](https://productname--itwhirlpool.myvtex.com/prodotti/lavaggio-e-asciugatura/lavatrici)

#### Screenshots

If the prop is set to true, then the behaviour is as shown in the screenshot below. 
![image](https://user-images.githubusercontent.com/73878310/120643715-ca0d8200-c476-11eb-9460-d067f910ba21.png)

If not set, or set to false, then the behaviour is the standard one.
![image](https://user-images.githubusercontent.com/73878310/120644039-2c668280-c477-11eb-96d8-ab74397ed41d.png)

##### Version
Version: 3.143.1
Proposed version: 3.144.0

victorhmp commented on docs/ProductName.md:
```diff
@@ -43,6 +43,7 @@ The `ProductName` is a block responsible for **displaying the product name** alo
 | `showProductReference` | `Boolean` | Show product reference | `false`| 
 | `showBrandName` | `Boolean` | Show brand name | `false`| 
 | `classes` | `CustomCSSClasses` | Used to override default CSS handles. To better understand how this prop works, we recommend reading about it [here](https://github.com/vtex-apps/css-handles#usecustomclasses). Note that this is only useful if you're using this block as a React component. | `undefined` |
+| `showProductLink` | `Boolean` | Show the product link associated with the its name. | `false`| 
```
Hi @anto90fg !
To keep our component's APIs as flexible as possible, without constantly making them bigger, we tend to stay away from boolean props.

In this case, to implement the functionality you want, how about creating a prop called `displayMode` (or something along those lines), which would accept the values `plain-text` and `link-to-product-page`? This way, if we need to add support to a new variation of this in the future, we just need to make the same prop accept one more value, instead of creating a whole new prop.

I'm not sold on these names I;'ve suggested 😅  Maybe @icazevedo has a better idea for them

icazevedo replied:
I like the names, but I think they should be camel-cased. `plainText` and `linkToProductPage`.

anto90fg replied:
I fixed it, following the requested changes. You can find all changes in the "required changes" commit.

icazevedo replied:
Please update the docs to reflect these changes 🙏

icazevedo commented on react/ProductName.tsx:
```diff
@@ -102,6 +109,37 @@ function ProductName({
+      <Wrapper
+        className={`${handles.productNameContainer} mv0 ${className ?? ''}`}
+      >
+        <a
```
I don't think this is the better way of doing it. We usually use `render-runtime`'s `Link` component to redirect users to the product page. Take a look at this [example](https://github.com/vtex-apps/product-summary/blob/46ade16bc8715835b39281cacec32f4823179b87/react/ProductSummaryCustom.tsx#L140).

I also think we shouldn't duplicate all the code in case there's a link wrap or not. You should create something like:

```suggestion
        const LinkWrapper = showProductLink ? Link : Fragment
        
        // And then use it where you put the a tag previously
        <LinkWrapper {...linkProps}>
          // Rest of the code
        </LinkWrapper>
```

anto90fg replied:
I fixed it, following the requested changes. You can find all changes in the "required changes" commit.

icazevedo reviewed (changes requested):
Thanks for the PR! I've left some required changes below.

victorhmp commented on react/ProductName.tsx:
```diff
@@ -138,6 +164,8 @@ function ProductNameWrapper(props: Props) {
   }
 
   const { product, selectedItem } = valuesFromContext
+  const displayMode =
+    props.displayMode !== undefined ? props.displayMode : 'plainText'
```
You don't need this check here, you can just set a default value in the actual `ProductName` component.

I prefer having it there to do as little work as possible inside this `ProductNameWrapper` component. It makes this whole file and components easier to understand. Also, since you're just setting a default value by checking if something is `undefined`, you can just use JS in your favor and not have to do this check manually :)

victorhmp commented on react/ProductName.tsx:
```diff
@@ -102,27 +113,42 @@ function ProductName({
     )
   }
 
+  const LinkWrapper =
```
Switching between rendering a <Link/> or a `<Fragment/>` is a nice idea! But passing `className` and all of the necessary props for the `Link` even if we're rendering a `<Fragment />` is not a good side effect. 

And I'm pretty sure would also flood the browser's cursor with errors/warnings from React, since `Fragment`s should only receive either a `key` property or none.

Swapping out the `Fragment` for a `div` might be a partial solution, but it isn't great either, since a bunch of props would still be passed to it.

I suggest creating a simple functional component, in this same file, that simply decides what to render in the end, something like:

```js
const LinkWrapper = ({ displayMode, linkProps, className, children }) => {
  if (displayMode === 'plainText') {
    return (
      <Fragment>
        {children}
      </Fragment>
    )
  }

  return (
    <Link className={className} {..linkProps}>
     {children}
    <Link />
  )
}
```

anto90fg replied:
I completely agree with the creation of a functional component!

victorhmp commented on react/ProductName.tsx:
```diff
@@ -76,6 +84,9 @@ function ProductName({
   skuName,
   brandName,
   productReference,
+  displayMode,
```
Add a default value for this prop right here

anto90fg replied:
Done!

icazevedo commented on react/ProductName.tsx:
```diff
@@ -59,6 +61,12 @@ type Props = {
+  displayMode?: 'plainText' | 'linkToProductPage'
+  /** Product link */
+  productLink?: string
+  /** Product link */
```
```suggestion
  /** Product id */
```

anto90fg replied:
Fixed :)

icazevedo commented on CHANGELOG.md:
```diff
@@ -7,6 +7,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ## [Unreleased]
 
+### Added
+- `showProductLink` prop to show the link associated with the product name.
```
Fix the prop name here too

anto90fg replied:
Fixed :)

camilavcoutinho commented on messages/context.json:
```diff
@@ -7,6 +7,7 @@
   "admin/editor.productName.showBrandName.title": "admin/editor.productName.showBrandName.title",
   "admin/editor.productName.showSku.title": "admin/editor.productName.showSku.title",
   "admin/editor.productName.showProductReference.title": "admin/editor.productName.showProductReference.title",
+  "admin/editor.productName.showProductLink.title": "Showing the link associated with the product name",
```
```suggestion
  "admin/editor.productName.showProductLink.title": "admin/editor.productName.showProductLink.title",
```

camilavcoutinho reviewed (changes requested):
For EN/PT/ES all good, the only thing that I think needs changes is context.json.

igorbrasileiro commented on react/ProductName.tsx:
```diff
@@ -102,31 +113,60 @@ function ProductName({
+/**
+ * Shows the link associated with the product name or plain text depending on the displayMode.
+ */
+const LinkWrapper = ({ displayMode, linkProps, className, children }: any) => {
```
Can you type LinkWrapper props instead of any? Would be nice for readability and typecheck

anto90fg replied:
Done! :)

igorbrasileiro commented on react/ProductName.tsx:
```diff
@@ -102,31 +113,60 @@ function ProductName({
+ * Shows the link associated with the product name or plain text depending on the displayMode.
+ */
+const LinkWrapper = ({ displayMode, linkProps, className, children }: any) => {
+  if (displayMode === 'plainText') {
+    return <Fragment>{children}</Fragment>
+  }
```
If you type props properly, you don't need to use this Fragment component, probably.

anto90fg replied:
How I could avoid using the fragment? I think it's a good choice to use it. But if you have a suggestion on how to remove it, I welcome it. :)

icazevedo replied:
I think you could just `return children` but I also think this is a matter of style. I see no problem with it staying as it is

anto90fg commented on messages/context.json:
```diff
@@ -7,6 +7,7 @@
   "admin/editor.productName.showBrandName.title": "admin/editor.productName.showBrandName.title",
   "admin/editor.productName.showSku.title": "admin/editor.productName.showSku.title",
   "admin/editor.productName.showProductReference.title": "admin/editor.productName.showProductReference.title",
+  "admin/editor.productName.showProductLink.title": "Showing the link associated with the product name",
```
Done :)

icazevedo reviewed (approved):
LGTM

victorhmp commented:
Hi @anto90fg !
I tried to review this again and it looks like your workspace is not working anymore :(
The one thing I was trying to check was how this shows up in the Site Editor. I'm pretty sure that the values you set in `enumNames` (used in this component's schema) is how each option shows up in the Site Editor. And as it stands, it would result in a bad UX for Site Editor users. Please use a translatable message for the `enumNames`, so that they can be more intuitive and also for i18n support.

anto90fg commented:
> Hi @anto90fg !
> I tried to review this again and it looks like your workspace is not working anymore :(
> The one thing I was trying to check was how this shows up in the Site Editor. I'm pretty sure that the values you set in `enumNames` (used in this component's schema) is how each option shows up in the Site Editor. And as it stands, it would result in a bad UX for Site Editor users. Please use a translatable message for the `enumNames`, so that they can be more intuitive and also for i18n support.

Hi @victorhmp , I inserted the enum name messages and moved the LinkWrapper component above its call, since the previous definition generated a lint error (the component was called before its definition).
I also updated the workspace ;)

igorbrasileiro commented:
The code LGTM

igorbrasileiro commented:
@anto90fg can you resolve the conflicts?

anto90fg commented:
> @anto90fg can you resolve the conflicts?

Done!
